File: java/06_061.java
import java.util.ArrayList;
import java.util.List;

public class NQueens {
   /**
    * Solve N-Queens problem using backtracking
    */
   public static int[][] solveNQueens(int n) {
      // Initialize board
      int[][] board = new int[n][n];
      
      if (!solveUtil(board, 0, n)) {
         return null;
      }
      
      return board;
   }

   /**
    * Check if position is safe for queen placement
    */
   private static boolean isSafe(int[][] board, int row, 
         int col, int n) {
      // Check row on left side
      for (int j = 0; j < col; j++) {
         if (board[row][j] == 1) {
            return false;
         }
      }
      
      // Check upper diagonal on left side
      for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
         if (board[i][j] == 1) {
            return false;
         }
      }
      
      // Check lower diagonal on left side
      for (int i = row, j = col; i < n && j >= 0; i++, j--) {
         if (board[i][j] == 1) {
            return false;
         }
      }
      
      return true;
   }

   /**
    * Recursive backtracking function
    */
   private static boolean solveUtil(int[][] board, int col, 
         int n) {
      // Base case: all queens placed
      if (col >= n) {
         return true;
      }

      // Try placing queen in each row of current column
      for (int i = 0; i < n; i++) {
         if (isSafe(board, i, col, n)) {
            // Place queen
            board[i][col] = 1;

            // Recur for next column
            if (solveUtil(board, col + 1, n)) {
               return true;
            }
            
            // Backtrack if no solution found
            board[i][col] = 0;
         }
      }
      
      return false;
   }

   /**
    * Format board for visualization
    */
   public static List<String> formatBoard(int[][] board) {
      if (board == null) {
         return null;
      }
      
      List<String> formatted = new ArrayList<>();
      for (int[] row : board) {
         StringBuilder sb = new StringBuilder();
         for (int cell : row) {
            sb.append(cell == 1 ? 'Q' : '.');
         }
         formatted.add(sb.toString());
      }
      return formatted;
   }

   /**
    * Visualize board solution as string
    */
   public static String visualizeSolution(int[][] board) {
      if (board == null) {
         return "No solution exists";
      }
      
      List<String> formatted = formatBoard(board);
      return String.join("\n", formatted);
   }

   /**
    * Main method for testing
    */
   public static void main(String[] args) {
      System.out.println("Testing different board sizes:");
      for (int n : new int[]{1, 2, 3, 4, 8}) {
         System.out.println("\nTesting " + n + "x" + n + 
               " board:");
         int[][] solution = solveNQueens(n);
         System.out.println(visualizeSolution(solution));
      }
   }
}

import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.*;
import java.util.List;

/**
 * Test class for NQueens solution
 */
public class TestNQueens {
   private NQueens solver;

   @Before
   public void setUp() {
      solver = new NQueens();
   }

   /**
    * Test 1x1 board (should have solution)
    */
   @Test
   public void test1x1Board() {
      int[][] result = NQueens.solveNQueens(1);
      assertNotNull(result);
      assertEquals(1, result.length);
      assertEquals(1, result[0][0]);
   }

   /**
    * Test 2x2 board (no solution)
    */
   @Test
   public void test2x2Board() {
      int[][] result = NQueens.solveNQueens(2);
      assertNull(result);
   }

   /**
    * Test 3x3 board (no solution)
    */
   @Test
   public void test3x3Board() {
      int[][] result = NQueens.solveNQueens(3);
      assertNull(result);
   }

   /**
    * Test 4x4 board (should have solution)
    */
   @Test
   public void test4x4Board() {
      int[][] result = NQueens.solveNQueens(4);
      assertNotNull(result);
      List<String> formatted = NQueens.formatBoard(result);
      assertEquals(4, formatted.size());
      
      // Count queens to ensure there are exactly 4
      int queenCount = 0;
      for (String row : formatted) {
         queenCount += row.chars().filter(ch -> ch == 'Q')
               .count();
      }
      assertEquals(4, queenCount);
   }

   /**
    * Test 8x8 board (should have solution)
    */
   @Test
   public void test8x8Board() {
      int[][] result = NQueens.solveNQueens(8);
      assertNotNull(result);
      List<String> formatted = NQueens.formatBoard(result);
      assertEquals(8, formatted.size());
      
      // Count queens to ensure there are exactly 8
      int queenCount = 0;
      for (String row : formatted) {
         queenCount += row.chars().filter(ch -> ch == 'Q')
               .count();
      }
      assertEquals(8, queenCount);
   }

   /**
    * Test no queens threaten each other
    */
   @Test
   public void testQueenThreats() {
      int n = 4;
      int[][] result = NQueens.solveNQueens(n);
      assertNotNull(result);
      
      // Check row threats (one queen per row)
      for (int[] row : result) {
         int sum = 0;
         for (int cell : row) {
            sum += cell;
         }
         assertEquals(1, sum);
      }
      
      // Check column threats (one queen per column)
      for (int j = 0; j < n; j++) {
         int sum = 0;
         for (int i = 0; i < n; i++) {
            sum += result[i][j];
         }
         assertEquals(1, sum);
      }
      
      // Check diagonal threats
      for (int i = 0; i < n; i++) {
         for (int j = 0; j < n; j++) {
            if (result[i][j] == 1) {
               // Check diagonals from each queen's position
               for (int k = 1; k < n; k++) {
                  // Check upper-right diagonal
                  if (i-k >= 0 && j+k < n) {
                     assertEquals(0, result[i-k][j+k]);
                  }
                  // Check lower-right diagonal
                  if (i+k < n && j+k < n) {
                     assertEquals(0, result[i+k][j+k]);
                  }
               }
            }
         }
      }
   }
}

File: python/06_061.py
import unittest

class NQueens:
   @staticmethod
   def solve_n_queens(n):
      def is_safe(board, row, col):
         # Check row on left side
         for j in range(col):
            if board[row][j] == 1:
               return False
               
         # Check upper diagonal on left side
         for i, j in zip(range(row, -1, -1), 
                        range(col, -1, -1)):
            if board[i][j] == 1:
               return False
               
         # Check lower diagonal on left side
         for i, j in zip(range(row, n, 1), 
                        range(col, -1, -1)):
            if board[i][j] == 1:
               return False
               
         return True

      def solve_util(board, col):
         # Base case: all queens placed
         if col >= n:
            return True

         for i in range(n):
            # Check if queen can be placed
            if is_safe(board, i, col):
               # Place the queen
               board[i][col] = 1

               # Recur to place rest of the queens
               if solve_util(board, col + 1):
                  return True
               
               # Remove queen if no solution found
               board[i][col] = 0
         
         # No solution with current configuration
         return False

      # Initialize board
      board = [[0 for x in range(n)] for y in range(n)]
      
      if not solve_util(board, 0):
         return None
         
      return board

   @staticmethod
   def format_board(board):
      # Return None if no solution
      if board is None:
         return None
      # Convert board to readable format
      return [''.join(['Q' if cell == 1 else '.' 
                     for cell in row]) 
             for row in board]

class TestNQueens(unittest.TestCase):
   def setUp(self):
      # Initialize solver
      self.solver = NQueens()

   def test_1x1_board(self):
      """Test case for 1x1 board - should have one solution"""
      result = self.solver.solve_n_queens(1)
      self.assertIsNotNone(result)
      self.assertEqual(len(result), 1)
      self.assertEqual(result[0][0], 1)

   def test_2x2_board(self):
      """Test case for 2x2 board - should have no solution"""
      result = self.solver.solve_n_queens(2)
      self.assertIsNone(result)

   def test_3x3_board(self):
      """Test case for 3x3 board - should have no solution"""
      result = self.solver.solve_n_queens(3)
      self.assertIsNone(result)

   def test_4x4_board(self):
      """Test case for 4x4 board - should have a solution"""
      result = self.solver.solve_n_queens(4)
      self.assertIsNotNone(result)
      formatted = self.solver.format_board(result)
      # One of the possible solutions for 4x4
      self.assertEqual(len(formatted), 4)
      self.assertEqual(sum(row.count('Q') 
                         for row in formatted), 4)

   def test_8x8_board(self):
      """Test case for 8x8 board - should have a solution"""
      result = self.solver.solve_n_queens(8)
      self.assertIsNotNone(result)
      formatted = self.solver.format_board(result)
      self.assertEqual(len(formatted), 8)
      self.assertEqual(sum(row.count('Q') 
                         for row in formatted), 8)

   def test_queen_threats(self):
      """Test that no queens threaten each other"""
      n = 4
      result = self.solver.solve_n_queens(n)
      self.assertIsNotNone(result)
      
      # Check row threats
      for row in result:
         self.assertEqual(sum(row), 1)
      
      # Check column threats
      for j in range(n):
         self.assertEqual(sum(result[i][j] 
                           for i in range(n)), 1)
      
      # Check diagonal threats
      for i in range(n):
         for j in range(n):
            if result[i][j] == 1:
               # Check diagonals from queen's position
               for k in range(1, n):
                  # Check upper-right diagonal
                  if i-k >= 0 and j+k < n:
                     self.assertEqual(result[i-k][j+k], 0)
                  # Check lower-right diagonal
                  if i+k < n and j+k < n:
                     self.assertEqual(result[i+k][j+k], 0)

def visualize_solution(board):
   """Helper function to visualize board solution"""
   if board is None:
      return "No solution exists"
   
   # Format the board
   formatted = NQueens.format_board(board)
   return '\n'.join(formatted)

if __name__ == '__main__':
   # Example usage and visualization
   print("Testing different board sizes:")
   for n in [1, 2, 3, 4, 8]:
      print(f"\nTesting {n}x{n} board:")
      solution = NQueens.solve_n_queens(n)
      print(visualize_solution(solution))
   
   # Run the unit tests
   unittest.main(argv=['first-arg-is-ignored'], 
                exit=False)